core/src/cluster_info.rs

```diff
@@ -544,27 +511,47 @@ impl ClusterInfo {
         stakes_and_index
     }
 
-    /// Return sorted Retransmit peers and index of `Self.id()` as if it were in that list
-    pub fn shuffle_peers_and_index<S: std::hash::BuildHasher>(
-        &self,
-        stakes: Option<&HashMap<Pubkey, u64, S>>,
+    fn stake_weighted_shuffle(
+        stakes_and_index: &[(u64, usize)],
         rng: ChaChaRng,
-    ) -> (usize, Vec<ContactInfo>) {
+    ) -> Vec<(u64, usize)> {
+        let stake_weights = stakes_and_index.iter().map(|(w, _)| *w).collect();
+
+        let shuffle = weighted_shuffle(stake_weights, rng);
+
+        shuffle.iter().map(|x| stakes_and_index[*x]).collect()
+    }
+
+    // Return sorted_retransmit_peers(including self) and their stakes
+    pub fn sorted_retransmit_peers_and_stakes(
+        &self,
+        stakes: Option<&HashMap<Pubkey, u64>>,
+    ) -> (Vec<ContactInfo>, Vec<(u64, usize)>) {
         let mut peers = self.retransmit_peers();
+        // insert "self" into this list for the layer and neighborhood computation
         peers.push(self.lookup(&self.id()).unwrap().clone());
-        let contacts_and_stakes: Vec<_> = ClusterInfo::stake_weighted_shuffle(&peers, stakes, rng);
-        let mut index = 0;
-        let peers: Vec<_> = contacts_and_stakes
-            .into_iter()
+        let stakes_and_index = ClusterInfo::sorted_stakes_with_index(&peers, stakes);
```

Comment:
Can you just save the self index from here to avoid the linear lookup in shuffle_peers_and_index?

Reply:
but self index changes with the weighted shuffle thats why I need to do that linear search